dnf/base.py

```diff
@@ -1363,26 +1386,41 @@ def cond_check(pkg):
                 pkg.requires in (pkg.name for pkg in self._goal._installs) or \
                 pkg.requires in [pkg.name for pkg in trans.install]
 
-        attr_fn = ((trans.install,
-                    functools.partial(self._goal.install, optional=True)),
-                   (trans.upgrade, self._goal.upgrade),
-                   (trans.remove,
-                    functools.partial(self._goal.erase, clean_deps=clean_deps)))
+        def trans_upgrade(query, remove_query):
+            sltr = dnf.selector.Selector(self.sack)
+            sltr.set(pkg=query)
+            self._goal.upgrade(select=sltr)
+            return remove_query
+
+        def trans_install(query, remove_query):
+            sltr = dnf.selector.Selector(self.sack)
+            sltr.set(pkg=query)
+            self._goal.install(select=sltr, optional=True)
+            return remove_query
+
+        def trans_remove(query, remove_query):
+            remove_query = remove_query.union(query)
+            return remove_query
+
+        remove_query = self.sack.query().filter(empty=True)
+        attr_fn = ((trans.install, trans_install),
+                   (trans.upgrade, trans_upgrade),
+                   (trans.remove, trans_remove))
 
         for (attr, fn) in attr_fn:
             for pkg in attr:
                 query_args = {'name': pkg.name}
                 if (pkg.basearchonly):
                     query_args.update({'arch': basearch})
-                q = self.sack.query().filter(**query_args).run()
```

Comment:
I guess that .apply() is used automatically in line bellow , but adding .apply() will not harm.